gufe/protocols/results.py

```diff
@@ -0,0 +1,46 @@
+# This code is part of OpenFE and is licensed under the MIT license.
+# For details, see https://github.com/OpenFreeEnergy/gufe
+
+import abc
+import uuid
+from typing import List, Optional, Any
+
+from pydantic import BaseModel, PrivateAttr
+
+
+class ProtocolUnitResult(BaseModel):
+    """Result for a single `ProtocolUnit` execution.
+
+    """
+    class Config:
+        extra = "allow"
+        
+    _dependencies: List[str] = PrivateAttr()
+    _uuid: str = PrivateAttr()
+
+    name: str  # name of the `ProtocolUnit` that produced this `ProtocolUnitResult`
+    data: Any  # should likely be fleshed out, currently a free-for-all
+
+    def __init__(self, 
+            dependencies: Optional[List["ProtocolUnitResult"]] = None,
+            **data):
+
+        super().__init__(**data)
+
+        self._uuid = str(uuid.uuid4())
```

Comment:
So in a world where we know the protocol id, run id, gen id etc, do we really need uuids?  Because yes they're nice for generating a unique thing, but they're (by design) awful to interpret, and we have enough knowledge to generate a useful identifier.

Reply:
Yes, but since we want to be able to deduplicate `ProtocolUnit`s across `ProtocolDAG`s within a `Scheduler`, it makes less sense generally to speak in terms of the source transformation that it came from, since there may be many such transformations that generated that particular `ProtocolUnit`. Giving the `ProtocolUnitResult`a unique "stamp" allows for unambiguous reference to its `_depdendencies`, which also have their own stamp, even once serialized and shipped to a `ResultStore`.

I do agree that generally UUIDs are to be avoided where possible, but in this case I think they are the least bad option. Even a hash of `ProtocolUnit.data` may not be that reasonable, since it may include references to rather large files and so may not even be a complete, differentiating hash across instances.